File: manga_frontend/src/services/list.service.ts
import { api, compactParams } from "./api";
import type { PaginatedResponse, UUID } from "@/types/common";
import type {
  ListCreatePayload,
  ListUpdatePayload,
  MangaListCollection,
  MangaListDetail,
  PublicListItem,
} from "@/types/list";

export const listService = {
  async mine(mangaId?: UUID) {
    const { data } = await api.get<MangaListCollection>("/lists/", {
      params: compactParams({ manga_id: mangaId }),
    });
    return data;
  },

  async create(payload: ListCreatePayload) {
    const { data } = await api.post<{ id: UUID; slug: string }>("/lists/", payload);
    return data;
  },

  async publicLists(params?: { page?: number; limit?: number; sort?: string; q?: string }) {
    const { data } = await api.get<PaginatedResponse<PublicListItem>>("/lists/public", {
      params: compactParams(params),
    });
    return data;
  },

  async detail(listId: UUID) {
    const { data } = await api.get<MangaListDetail>(`/lists/${listId}`);
    return data;
  },

  async update(listId: UUID, payload: ListUpdatePayload) {
    const { data } = await api.put<{ success: boolean }>(`/lists/${listId}`, payload);
    return data;
  },

  async remove(listId: UUID) {
    await api.delete(`/lists/${listId}`);
  },

  async addItem(listId: UUID, mangaId: UUID) {
    const { data } = await api.post<{ success?: boolean; message?: string; item_count: number }>(
      `/lists/${listId}/items`,
      null,
      { params: { manga_id: mangaId } },
    );
    return data;
  },

  async removeItem(listId: UUID, mangaId: UUID) {
    const { data } = await api.delete<{ success: boolean; item_count: number }>(
      `/lists/${listId}/items/${mangaId}`,
    );
    return data;
  },

  async follow(listId: UUID) {
    const { data } = await api.post<{ success?: boolean; message?: string; follower_count?: number }>(
      `/lists/${listId}/follow`,
    );
    return data;
  },

  async unfollow(listId: UUID) {
    const { data } = await api.delete<{ success?: boolean; message?: string; follower_count?: number }>(
      `/lists/${listId}/follow`,
    );
    return data;
  },
};

File: ktgk_2324802010044/lib/services/list_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../core/dio_client.dart';
import '../models/manga_list.dart';

class ListService {
  static final _dio = DioClient.instance;

  /// GET /lists/?manga_id={id} → trả {my_lists: [...], followed_lists: [...]}
  static Future<Map<String, List<MangaListBrief>>> getMyLists({
    String? mangaId,
  }) async {
    try {
      final params = <String, dynamic>{};
      if (mangaId != null) params['manga_id'] = mangaId;

      final response = await _dio.get('/lists/', queryParameters: params);
      final data = response.data as Map<String, dynamic>;

      final myListsJson = data['my_lists'] as List? ?? [];
      final followedListsJson = data['followed_lists'] as List? ?? [];

      return {
        'my_lists': myListsJson
            .map((e) => MangaListBrief.fromJson(e as Map<String, dynamic>))
            .toList(),
        'followed_lists': followedListsJson
            .map((e) => MangaListBrief.fromJson(e as Map<String, dynamic>))
            .toList(),
      };
    } on DioException catch (e) {
      debugPrint('ListService.getMyLists error: ${e.response?.data}');
      return {'my_lists': [], 'followed_lists': []};
    }
  }

  /// POST /lists/  body: {Name, Description?, Visibility}
  static Future<Map<String, dynamic>> createList(
    String name, {
    String? description,
    String visibility = 'private',
  }) async {
    try {
      final response = await _dio.post(
        '/lists/',
        data: {
          'Name': name,
          'Description': description,
          'Visibility': visibility,
        },
      );
      return response.data as Map<String, dynamic>;
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to create list');
    }
  }

  /// PUT /lists/{id}  body: {Name?, Description?, Visibility?}
  static Future<void> updateList(
    String listId, {
    String? name,
    String? description,
    String? visibility,
  }) async {
    try {
      final data = <String, dynamic>{};
      if (name != null) data['Name'] = name;
      if (description != null) data['Description'] = description;
      if (visibility != null) data['Visibility'] = visibility;

      await _dio.put('/lists/$listId', data: data);
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to update list');
    }
  }

  /// DELETE /lists/{id}
  static Future<void> deleteList(String listId) async {
    try {
      await _dio.delete('/lists/$listId');
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to delete list');
    }
  }

  /// GET /lists/{id} → ListDetailResponse
  static Future<MangaListDetail> getListDetail(String listId) async {
    try {
      final response = await _dio.get('/lists/$listId');
      return MangaListDetail.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      throw Exception(
        e.response?.data?['detail'] ?? 'Failed to get list detail',
      );
    }
  }

  /// POST /lists/{id}/items?manga_id={mangaId}
  static Future<void> addItem(String listId, String mangaId) async {
    try {
      await _dio.post(
        '/lists/$listId/items',
        queryParameters: {'manga_id': mangaId},
      );
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to add item');
    }
  }

  /// DELETE /lists/{id}/items/{mangaId}
  static Future<void> removeItem(String listId, String mangaId) async {
    try {
      await _dio.delete('/lists/$listId/items/$mangaId');
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to remove item');
    }
  }

  /// POST /lists/{id}/follow
  static Future<void> followList(String listId) async {
    try {
      await _dio.post('/lists/$listId/follow');
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to follow list');
    }
  }

  /// DELETE /lists/{id}/follow
  static Future<void> unfollowList(String listId) async {
    try {
      await _dio.delete('/lists/$listId/follow');
    } on DioException catch (e) {
      throw Exception(e.response?.data?['detail'] ?? 'Failed to unfollow list');
    }
  }

  /// GET /lists/public?page=&q=&sort=
  static Future<Map<String, dynamic>> getPublicLists({
    int page = 1,
    int limit = 12,
    String? query,
    String sort = 'updated_desc',
  }) async {
    try {
      final params = <String, dynamic>{
        'page': page,
        'limit': limit,
        'sort': sort,
      };
      if (query != null && query.isNotEmpty) params['q'] = query;

      final response = await _dio.get('/lists/public', queryParameters: params);
      final data = response.data as Map<String, dynamic>;
      if (data['items'] is List) return data;
      if (data['lists'] is List) {
        return {
          ...data,
          'items': data['lists'],
          'total': data['total'] ?? (data['lists'] as List).length,
        };
      }
      return {'items': [], 'total': 0};
    } on DioException catch (e) {
      debugPrint('ListService.getPublicLists error: ${e.response?.data}');
      return {'items': [], 'total': 0};
    }
  }
}
